searx/sqlitedb.py

```diff
@@ -0,0 +1,322 @@
+        """
+        if sys.version_info < (3, 12):
+            # Prior Python 3.12 there is no "autocommit" option
+            self.SQLITE_CONNECT_ARGS.pop("autocommit", None)
+
+        self.init()
+        logger.debug("%s: connect to DB: %s // %s", self.__class__.__name__, self.db_url, self.SQLITE_CONNECT_ARGS)
+        conn = sqlite3.Connection(self.db_url, **self.SQLITE_CONNECT_ARGS)  # type: ignore
+        conn.execute(f"PRAGMA journal_mode={self.SQLITE_JOURNAL_MODE}")
+        self.register_functions(conn)
+        return conn
+
+    def register_functions(self, conn):
+        """Create user-defined_ SQL functions.
+
+        ``REGEXP(<pattern>, <field>)`` : 0 | 1
+           `re.search`_ returns (int) 1 for a match and 0 for none match of
+           ``<pattern>`` in ``<field>``.
+
+           .. code:: sql
+
+              SELECT '12' AS field WHERE REGEXP('^[0-9][0-9]$', field)
+              -- 12
+
+              SELECT REGEXP('[0-9][0-9]', 'X12Y')
+              -- 1
+              SELECT REGEXP('[0-9][0-9]', 'X1Y')
+              -- 0
+
+        .. _user-defined: https://docs.python.org/3/library/sqlite3.html#sqlite3.Connection.create_function
+        .. _deterministic: https://sqlite.org/deterministic.html
+        .. _re.search: https://docs.python.org/3/library/re.html#re.search
+        """
+
+        conn.create_function('regexp', 2, lambda x, y: 1 if re.search(x, y) else 0, deterministic=True)
+
+    @property
+    def DB(self) -> sqlite3.Connection:
+        """Provides a DB connection.  The connection is a *singleton* and
+        therefore well suited for read access.  If
+        :py:obj:`SQLITE_THREADING_MODE` is ``serialized`` only one DB connection
+        is created for all threads.
+
+        .. note::
+
+           For dedicated `transaction control`_, it is recommended to create a
+           new connection (:py:obj:`SQLiteAppl.connect`).
+
+        .. _transaction control:
+            https://docs.python.org/3/library/sqlite3.html#sqlite3-controlling-transactions
+        """
+
+        if getattr(self.thread_local, 'DB', None) is None:
+            self.thread_local.DB = self.connect()
+
+        # Theoretically it is possible to reuse the DB cursor across threads as
+        # of Python 3.12, in practice the threading of the cursor seems to me to
+        # be so faulty that I prefer to establish one connection per thread
+
+        return self.thread_local.DB
+
+        # In "serialized" mode, SQLite can be safely used by multiple threads
+        # with no restriction.
+        #
+        # if self.SQLITE_THREADING_MODE != "serialized":
+        #     if getattr(self.thread_local, 'DB', None) is None:
+        #         self.thread_local.DB = self.connect()
+        #     return self.thread_local.DB
+        #
+        # if self._DB is None:
+        #     self._DB = self.connect()  # pylint: disable=attribute-defined-outside-init
+        # return self._DB
```

Comment:
I had a lot of trouble with the optimization of DB connections in Python 3.12 ..  https://docs.python.org/3/library/sqlite3.html#sqlite3.threadsafety says

> 3. Serialized: In serialized mode, SQLite can be safely used by multiple threads with no restriction.
> .. Threads may share the module, connections and cursors

But I have seen lots of issues when sharing the DB cursor over multiple threads.

to avoid this problem, one DB connection per thread is now set up for “read-only accesses” / a new connection had to be created for SQL updates, as the SQL updates in the different threads would otherwise block each other (as long as a transaction is running in one thread, the other thread may not make any updates, these competing updates could only be synchronized with separate DB connections)